units/en/unit2/hands-on.mdx

```diff
@@ -0,0 +1,1074 @@
+    "eval_seed": eval_seed,
+    "learning_rate": learning_rate,
+    "gamma": gamma,
+    "max_epsilon": max_epsilon,
+    "min_epsilon": min_epsilon,
+    "decay_rate": decay_rate,
+    "qtable": Qtable_frozenlake,
+}
+```
+
+Let's fill the `package_to_hub` function:
+
+- `repo_id`: the name of the Hugging Face Hub Repository that will be created/updated `
+(repo_id = {username}/{repo_name})`
+💡 A good `repo_id` is `{username}/q-{env_id}`
+- `model`: our model dictionary containing the hyperparameters and the Qtable.
+- `env`: the environment.
+- `commit_message`: message of the commit
+
+```python
+model
+```
+
+```python
+username = ""  # FILL THIS
+repo_name = "q-FrozenLake-v1-4x4-noSlippery"
+push_to_hub(repo_id=f"username}/{repo_name}", model=model, env=env)
+```
+
+Congrats 🥳 you've just implemented from scratch, trained and uploaded your first Reinforcement Learning agent.
+FrozenLake-v1 no_slippery is very simple environment, let's try an harder one 🔥.
+
+# Part 2: Taxi-v3 🚖
+
+## Create and understand [Taxi-v3 🚕](https://www.gymlibrary.dev/environments/toy_text/taxi/)
+---
+
+💡 A good habit when you start to use an environment is to check its documentation
+
+👉 https://www.gymlibrary.dev/environments/toy_text/taxi/
+
+---
+
+In `Taxi-v3` 🚕, there are four designated locations in the grid world indicated by R(ed), G(reen), Y(ellow), and B(lue).
+
+When the episode starts, **the taxi starts off at a random square** and the passenger is at a random location. The taxi drives to the passenger’s location, **picks up the passenger**, drives to the passenger’s destination (another one of the four specified locations), and then **drops off the passenger**. Once the passenger is dropped off, the episode ends.
+
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/notebooks/unit2/taxi.png" alt="Taxi">
+
+
+```python
+env = gym.make("Taxi-v3")
+```
+
+There are **500 discrete states since there are 25 taxi positions, 5 possible locations of the passenger** (including the case when the passenger is in the taxi), and **4 destination locations.**
+
+
+```python
+state_space = env.observation_space.n
+print("There are ", state_space, " possible states")
+```
+
+```python
+action_space = env.action_space.n
+print("There are ", action_space, " possible actions")
+```
+
+The action space (the set of possible actions the agent can take) is discrete with **6 actions available 🎮**:
+
+- 0: move south
+- 1: move north
+- 2: move east
+- 3: move west
+- 4: pickup passenger
+- 5: drop off passenger
+
+Reward function 💰:
+
+- -1 per step unless other reward is triggered.
+- +20 delivering passenger.
+- -10 executing “pickup” and “drop-off” actions illegally.
+
+```python
+# Create our Q table with state_size rows and action_size columns (500x6)
+Qtable_taxi = initialize_q_table(state_space, action_space)
+print(Qtable_taxi)
+print("Q-table shape: ", Qtable_taxi.shape)
+```
+
+## Define the hyperparameters ⚙️
```

Comment:
FYI: If you repeat headings, it won't work well in the navigation bar on the Hub (see right of https://moon-ci-docs.huggingface.co/deep-rl-course/pr_105/en/unit2/hands-on )